File: okone/src/main/java/com/cdh/okone/priority/PriorityArrayDeque.java
package com.cdh.okone.priority;

import java.util.ArrayDeque;
import java.util.Collection;
import java.util.Deque;
import java.util.Iterator;
import java.util.LinkedList;
import java.util.List;
import java.util.ListIterator;
import java.util.PriorityQueue;
import java.util.Queue;
import java.util.TreeMap;
import java.util.TreeSet;

/**
 * 优先级ArrayDeque
 * 从大到小排序
 * Created by chidehang on 2021/1/12
 */
public class PriorityArrayDeque<E> extends ArrayDeque<E> implements Deque<E> {

    private LinkedList<E> queue;

    public PriorityArrayDeque() {
        queue = new LinkedList<>();
    }

    public PriorityArrayDeque(int numElements) {
        super(numElements);
        queue = new LinkedList<>();
    }

    public PriorityArrayDeque(Collection<? extends E> c) {
        super(c);
        queue = new LinkedList<>(c);
        for (E e : c) {
            sortAdd(e);
        }
    }

    @Override
    public void addFirst(E e) {
        super.addFirst(e);
        sortAdd(e);
    }

    @Override
    public void addLast(E e) {
        super.addLast(e);
        sortAdd(e);
    }

    /**
     * 按序插入
     */
    private void sortAdd(E x) {
        Comparable<? super E> key = (Comparable<? super E>) x;
        ListIterator<E> i = queue.listIterator(queue.size());
        while (i.hasPrevious()) {
            E e = i.previous();
            if (key.compareTo(e) <= 0) {
                i.next();
                i.add(x);
                return;
            }
        }
        queue.addFirst(x);
    }

    @Override
    public E pollFirst() {
        queue.pollFirst();
        return super.pollFirst();
    }

    @Override
    public E pollLast() {
        queue.pollLast();
        return super.pollLast();
    }

    @Override
    public boolean removeFirstOccurrence(Object o) {
        queue.remove(o);
        return super.removeFirstOccurrence(o);
    }

    @Override
    public boolean removeLastOccurrence(Object o) {
        queue.remove(o);
        return super.removeLastOccurrence(o);
    }

    @Override
    public void clear() {
        super.clear();
        queue.clear();
    }

    @Override
    public Iterator<E> iterator() {
        return queue.iterator();
    }
}

File: okone/src/main/kotlin/com/cdh/okone/priority/PriorityArrayDeque.kt
package com.cdh.okone.priority

import java.util.*

/**
 * 优先级ArrayDeque
 * 从大到小排序
 * Created by chidehang on 2021/1/12
 */
class PriorityArrayDeque<E> : ArrayDeque<E>, Deque<E> {
    private var queue: LinkedList<E>

    constructor() {
        queue = LinkedList()
    }

    constructor(numElements: Int) : super(numElements) {
        queue = LinkedList()
    }

    constructor(c: Collection<E>) : super(c) {
        queue = LinkedList(c)
        for (e in c) {
            sortAdd(e)
        }
    }

    override fun addFirst(e: E) {
        super.addFirst(e)
        sortAdd(e)
    }

    override fun addLast(e: E) {
        super.addLast(e)
        sortAdd(e)
    }

    /**
     * 按序插入
     */
    private fun sortAdd(x: E) {
        val key = x as Comparable<E>
        val i = queue.listIterator(queue.size)
        while (i.hasPrevious()) {
            val e = i.previous()
            if (key <= e) {
                i.next()
                i.add(x)
                return
            }
        }
        queue.addFirst(x)
    }

    override fun pollFirst(): E? {
        queue.pollFirst()
        return super.pollFirst()
    }

    override fun pollLast(): E? {
        queue.pollLast()
        return super.pollLast()
    }

    override fun removeFirstOccurrence(o: Any?): Boolean {
        queue.remove(o)
        return super.removeFirstOccurrence(o)
    }

    override fun removeLastOccurrence(o: Any?): Boolean {
        queue.remove(o)
        return super.removeLastOccurrence(o)
    }

    override fun clear() {
        super.clear()
        queue.clear()
    }

    override fun iterator(): MutableIterator<E> {
        return queue.iterator()
    }
}